### DISCUSSION TITLE
Feed Plugin Error

### DISCUSSION THREAD

**Discussion by fdevans** [vuepress-theme-hope/vuepress-theme-hope#3480] [Q&A] [answered]
Getting this error when running a build with the Feed plugin V2:

```
error error in hook onGenerated from vuepress-plugin-feed2
Error: EISDIR: illegal operation on a directory, read
    at Object.readSync (node:fs:751:3)
    at tryReadSync (node:fs:451:20)
    at Object.readFileSync (node:fs:497:19)
    at Dt (file:///Users/forrestevans/Documents/GitHub/docs/node_modules/vuepress-plugin-feed2/lib/node/index.js:3:6687)
    at get content [as content] (file:///Users/forrestevans/Documents/GitHub/docs/node_modules/vuepress-plugin-feed2/lib/node/index.js:3:9332)
    at St.getFeedItem (file:///Users/forrestevans/Documents/GitHub/docs/node_modules/vuepress-plugin-feed2/lib/node/index.js:3:10326)
    at Xt.addPages (file:///Users/forrestevans/Documents/GitHub/docs/node_modules/vuepress-plugin-feed2/lib/node/index.js:3:13693)
    at file:///Users/forrestevans/Documents/GitHub/docs/node_modules/vuepress-plugin-feed2/lib/node/index.js:3:14064
    at Array.map (<anonymous>)
    at Xt.generateFeed (file:///Users/forrestevans/Documents/GitHub/docs/node_modules/vuepress-plugin-feed2/lib/node/index.js:3:13919)
    at Object.onGenerated [as hook] (file:///Users/forrestevans/Documents/GitHub/docs/node_modules/vuepress-plugin-feed2/lib/node/index.js:3:16202)
    at Object.process (file:///Users/forrestevans/Documents/GitHub/docs/node_modules/@vuepress/core/dist/index.js:681:37)
    at async CAC.<anonymous> (file:///Users/forrestevans/Documents/GitHub/docs/node_modules/@vuepress/cli/dist/index.js:234:3)

```

The config I'm using is:
```
        feedPlugin({
            hostname: 'https://docs.rundeck.com',
            rss: true,
            json: true
            sort: entries => _.reverse(_.sortBy(entries, 'date'))
        }),
```

Any guidance on what might be happening?

There is a branch on our public docs site with the full config here: https://github.com/rundeck/docs/tree/vuepress2-migration/docs/.vuepress

**Comment by fdevans** [comment, marked as answer]
I was able to figure this out.  It seems since some of my pages don't have any frontmatter or contained frontmatter without a 'feed' section the default filter was struggling.  I had to do some things in our vuepress V1 to V2 migration.

1. Migrate the config to Typescript.  We had been using a separate plugins file and importing it for cleanliness, but moving it all to the main `config.ts` helped.
2. I had to use the config below with the specified filter to collect only the files that had frontmatter with a feed section established.
3. Our previous feed plugin (not from this same author I don't think...) wanted dates as a string.  This expects them to be presented without quotes around them (so number)

Working Config to sort by json/rss feeds only by date, descending, only when a feed section exists in frontmatter:
```
feedPlugin({
        hostname: 'https://docs.rundeck.com',
        rss: true,
        json: true,
        filter: ({ frontmatter }: Page): boolean =>
            frontmatter.feed !== undefined,
        sort: ({ frontmatter }: Page): number => _.reverse(_.sortBy(entries, frontmatter.date))
    })

```
also had to include:
`import _ from 'lodash';`

Frontmatter date entries will be changed as shown below as part of content migration steps.

`date: "2023-10-03"`  to `date: 2023-10-03`

**Comment by Mister-Hope** [comment]
I don't think an empty frontmatter stops the generate process, ~~but yes, if your date settings are conflict eith the default date getter, the plugin may throw.~~
Checked the docs, seems that we don't have any sorter opt-in.

Can you add more detail how an fs problem is thrown? And whether ee should do some improvements?

**Reply by fdevans** [reply]
I can try to replicate it again, but we are upgrading from V1 to V2 of our site and there's a lot changing every day, along with myself learning a lot more about how it all works together.

At the point above I had it working, but now we've moved over to using the full hope theme and my filter isn't working the same anymore.  I moved the feed configuration to the Theme configuration and now it's pulling more pages than when it was just installed as a plugin directly.

This is the config we are using 
``` 
theme: hopeTheme({
    plugins: {
      feed: {
        hostname: 'https://docs.rundeck.com',
        rss: true,
        json: true,
        filter: ({ frontmatter }: Page): boolean => (frontmatter.feed !== undefined),
        sort: ({ frontmatter }: Page): number => _.reverse(_.sortBy(entries, frontmatter.date))
      },
}
})
```
It seems the added entries that I wasn't getting before are directories. (example)

```
    {
      "title": "Api",
      "url": "https://docs.rundeck.com/api/",
      "id": "https://docs.rundeck.com/api/",
      "content_html": "",
      "date_modified": "2023-10-03T18:10:18.228Z",
      "authors": [],
      "tags": []
    },
```

I will try to figure out how to pull only the pages with a feed section in frontmatter and exclude everything else.  Suggestions welcome :)

**Reply by fdevans** [reply]
Seems I just needed to include the defaults along with my additional filter entry. Setting the filter to the following fixed the issue and I'm now only getting the pages I expected.

`filter: ({ frontmatter, filePathRelative }: Page): boolean => !(frontmatter.feed === undefined || frontmatter.home || !filePathRelative || frontmatter.article === false || frontmatter.feed === false),`

The sort function isn't quite right yet, but I'll post that when I figure it out.

**Reply by fdevans** [reply]
We can't seem to figure out how to use the Sort function.  https://plugin-feed2.vuejs.press/config/#sort

If there are additional examples it would be helpful.  We have `frontmatter.date` populated with date values, but the sort is not coming out descending.

**Reply by fdevans** [reply]
Since it's a new question I'll open a new thread. This one was about the filter, which we figured out.
